### ISSUE TITLE
Add rule to detekt missing `else` for `when` on non-sealed type.

### ISSUE THREAD

**Issue by yogurtearl** [detekt/detekt#6688]
If I have this (any sealed type):
```kotlin
sealed interface Foo
class Bar: Foo
```

And we have this code, no problems so far, because it is exhaustive.
```kotlin
val foo: Foo = getFoo()
when(foo) {
    is Bar -> println("Bar")
}
```

But then later someone **removed the `sealed` from `Foo`** then we have: 
```kotlin
interface Foo
class Bar: Foo
```

```kotlin
val foo: Foo = getFoo()
when(foo) {
    is Bar -> println("Bar")
    /// missing an else here, anything but Bar is silently ignored.
}
```

Suggesting a rule for `MissingElseForWhenOnNonSealedType`, which requires an `else` on any `when(sealed)`/`is SubType` construct for any non-sealed `Type`, where not all the values are know at compile time.

Non-compliant:

```kotlin
interface Foo
class Bar: Foo

val foo: Foo = getFoo()
when(foo) {
    is Bar -> println("Bar")
    /// missing an else here anything but Bar is ignored.
}
```

Compliant: 
```kotlin
interface Foo
class Bar: Foo

val foo: Foo = getFoo()
when(foo) {
    is Bar -> println("Bar")
    else -> println(foo)
}
```

**Comment by BraisGabin** [comment]
I thought that we already had this rule. I'm at the mobile so I can't check it. But if we don't have it for sure we should add it.

**Comment by yogurtearl** [comment]
I had an instance of this that didn't trigger any rules and then I looked though the rules expecting to find one, but I didn't see a rule for this.

**Comment by yogurtearl** [comment]
I am using Detekt `1.23.4` with Kotlin 1.9.20 and Gradle 8.5, with type resolution turned on. 
The instance of this I was looking at was in an Andoid subproject.

**Comment by cortinico** [comment]
> I thought that we already had this rule. I'm at the mobile so I can't check it. But if we don't have it for sure we should add it.

We do have a similar rule which does a different check `ElseCaseInsteadOfExhaustiveWhen`
